pxtblocks/fields/field_melodySandbox.ts

```diff
@@ -0,0 +1,415 @@
+            Blockly.DropDownDiv.clearContent();
+            Blockly.DropDownDiv.setColour(this.getDropdownBackgroundColour(), this.getDropdownBorderColour());
+            let contentDiv = Blockly.DropDownDiv.getContentDiv() as HTMLDivElement;
+            goog.style.setStyle(contentDiv, "max-height", "500px");
+            this.renderEditor(Blockly.DropDownDiv.getContentDiv() as HTMLDivElement);
+            this.createGridDispaly();
+            Blockly.DropDownDiv.showPositionedByBlock(this, this.sourceBlock_, () => {
+                this.onEditorClose();
+                // revert all style attributes for dropdown div
+                goog.style.setStyle(contentDiv, "max-height", null);
+            });
+        }
+
+        getText() {
+            this.stringRep = this.getTypeScriptValue();
+            return this.stringRep;
+        }
+
+        setText(newText: string) {
+            if (newText == null || newText == "" || newText == "\"\"" || (this.stringRep && this.stringRep === newText)) { // ignore empty strings
+                return;
+            }
+            this.stringRep = newText;
+            this.parseTypeScriptValue(newText);
+        }
+
+
+        // This will be run when the field is created (i.e. when it appears on the workspace)
+        protected onInit() {
+            this.render_();
+            if (!this.melody) {
+                this.melody = new pxtmelody.MelodyArray();
+            }
+            Blockly.FieldLabel.prototype.setText.call(this, this.getTitle());
+        }
+
+        // Render the editor that will appear in the dropdown div when the user clicks on the field
+        protected renderEditor(div: HTMLDivElement) {
+            let topDiv = goog.dom.createDom("div", {}) as HTMLElement;
+            topDiv.className = "melody-top-bar-div";
+
+            let editorGalleryToggle = goog.dom.createDom("div", {}) as HTMLElement;
+            editorGalleryToggle.id = "melody-toggle";
+            let editorButton = document.createElement("button");
+            editorButton.innerText = "Editor";
+            editorButton.className = "ui left attached button";
+            let galleryButton = document.createElement("button");
+            galleryButton.innerText = "Gallery";
+            galleryButton.className = "right attached ui button";
+            editorGalleryToggle.appendChild(editorButton);
+            editorGalleryToggle.appendChild(galleryButton);
+
+            topDiv.appendChild(editorGalleryToggle);
+            let melodyName = document.createElement("input");
+            melodyName.className = "ui input";
+            melodyName.id = "melody-name";
+            melodyName.maxLength = 25;
+            melodyName.value = this.title;
+            melodyName.addEventListener("input", () => this.setTitle(melodyName.value));
+            topDiv.appendChild(melodyName);
+            div.appendChild(topDiv);
+
+            let gridDiv = goog.dom.createDom("div", {}) as HTMLElement;
```

Comment:
use document.createElement as above

Reply:
I've seen `goog.dom.createDom("div", {}) as HTMLElement;` in other files. When would I want to use one over the other?

Reply:
goog.dom is a Blockly dependency that they are trying to get rid of. At some point we'll need to fix it elsewhere, but for now we don't want to check in new code that uses it